### ISSUE TITLE
Optional map instead of docstring parsing

### ISSUE THREAD

**Issue by isysd** [gangverk/flask-swagger#18]
Transferred from: https://github.com/deginner/flask-swagger/issues/2

Some people prefer to leave Python docstrings for internal facing documentation. In these cases, a map should be allowed to match the route against a swagger path description.

Basically, if `map` is given to flask-swagger, then it is used in place of `_parse_docstring`.

**Comment by rmoorman** [comment]
I am in the camp "docstrings for (internal) python code documentation" too. Maybe a decorator for endpoints would be great addition (see https://github.com/rantav/flask-restful-swagger#how-to for an example how another project is using this approach for generating swagger api specs).
What a coincidence by the way. Maybe the projects could be joined :-) (as flask-restful-swagger is left without maintenance currently)?

**Comment by atlithorn** [comment]
I agree with the idea of providing an existing map - I assume the idea is to expand on the template idea that @isysd added, taking care not to override existing endpoints of the template with "internal docs". 
A fine idea.

Part of the reason I didn't enhance flask-restful-swagger when we started out (we were using them to begin with) was that we found editing the json in the decorators cumbersome so I'm not too keen to add that functionality.

**Comment by rmoorman** [comment]
@atlithorn how about adding a decorator and providing a string instead of a nested data structure?
What I meant is something like this:
~~~~.python
class UserAPI(MethodView):
    @swaggerspecdecorator("""
        Create a new user
        ---
        tags:
          - users
        # ....
    """)
    def post(self):
        """
        And the internal documentation goes here as usual
        """
        return {}
~~~~

**Comment by isysd** [comment]
I had initially thought to do some sort of mapping with the template, but it seems like a nominal increase in functionality. @rmoorman has a better strategy, keeping the swagger path documentation close to the route handler, but not in the actual docstring. This is obviously a big deviation from the current behavior, however, so I'm not sure what the best way to proceed is.

**Comment by rmoorman** [comment]
@isysd couldn't the `_parse_docstring` function be adjusted to check if a property exists on the method (`method._swagger_spec` or the like) which then would be parsed (maybe with `inspect.cleandoc`)  instead of calling the `inspect.getdoc` function [here](https://github.com/gangverk/flask-swagger/blob/master/flask_swagger.py#L22)? The decorator would then add the property and the current way of using the docstrings would still work I guess.

**Comment by atlithorn** [comment]
Understood. The flask-restful-swagger decoration approach entails keeping a registry in memory for the apps lifetime. 
Using the template addition you wrote users could write their own decorators to do whatever they please. Write docs to file or memory which the spec endpoint could provide as a template.

I'm not keen on making the decorator part of the library (at least sight unseen) but it would definitely be a cool addition to the examples

**Comment by rmoorman** [comment]
@atlithorn by the way, I am not saying here that the "whole thing done by flask-swagger" is great (didn't have a good look at it) it just seems that a decorator provides a cleaner way to add the swagger "docstring" to an api method than adding it to the normal method docstring IMHO. It provides a better separation of the documentation concerns at hand here I think.
Also, docstrings tend to use restructured text as far as I know. Wouldn't keeping the swagger spec inside the docstring interfere with tools like sphinx?
Anyway, it would be great if a decorator would be used for adding the swagger specification :-)

**Comment by atlithorn** [comment]
Your Sphinx point is valid.

I'm open to revisiting the decorator approach if we find a suitable solution

**Comment by rmoorman** [comment]
@atlithorn which issues with the decorator approach do you see and would you like to see addressed?

**Comment by atlithorn** [comment]
Sorry for the radio silence, I was on vacation. 
My problem with the decorator approach used by flask-restful-swagger was twofold.

1. It entails building a registry that the server needs to keep in memory for the duration of it's lifetime, which just felt wrong for a request that runs very seldomly. Flask-swagger only runs at request time. I will admit that this is probably just me being anal and it could easily be solved by allowing persistence to file instead of memory.
2. They used json which is much more problematic to both read and edit for the developer. Using YAML in the decorator instead is trivial.

**Comment by rochacbruno** [comment]
In Flasgger I am using @swag_from https://github.com/rochacbruno/flasgger#handling-multiple-http-methods-and-routes-for-a-single-function

**Comment by isysd** [comment]
Sorry to open up a can of worms with this issue and then disappear. In truth, someone on our team voiced this complaint but never followed up with a suggestion. We've overall been happy using flask-swagger with the docstring spec fragments. Still, it seems like a valid concern that maybe others also care about.
